language/e2e-testsuite/src/tests/preburn_queue.rs

```diff
@@ -0,0 +1,407 @@
+    cancel_burn_with_amount_new(&mut executor, 10, &tc_account, &mut tc_seqno, false);
+}
+
+/// Make sure we can preburn and before the upgrade with the old scripts, and that we can
+/// preburn and burn after the upgrade with the new scripts (and the old preburn script) and that
+/// concurrent preburns are supported immediately without any additional input from TC. Note that
+/// the preburn balance at the time of upgrade is zero.
+#[test]
+fn concurrent_preburns_test_upgrade_empty_preburn_multiple_initial_preburns() {
+    let (mut executor, dr_account, tc_account, dd_account) = start_with_released_df();
+    let mut tc_seqno = 0;
+    let mut dd_seqno = 0;
+    let mut dr_seqno = 1;
+
+    create_preburn_balance(&mut executor, &dd_account, 10, &mut dd_seqno, false);
+    burn_old(&mut executor, &tc_account, &mut tc_seqno, false);
+
+    create_preburn_balance(&mut executor, &dd_account, 10, &mut dd_seqno, false);
+    cancel_burn_old(&mut executor, &tc_account, &mut tc_seqno, false);
+
+    // Upgrade the DF
+    upgrade_df(&mut executor, &dr_account, &mut dr_seqno);
+
+    // Now make sure that we can upgrade the DD
+    create_preburn_balance(&mut executor, &dd_account, 100, &mut dd_seqno, false);
+    // Make sure we can now have multiple preburn requests
+    create_preburn_balance(&mut executor, &dd_account, 20, &mut dd_seqno, false);
+
+    // burn and cancel in a different order
+    burn_with_amount_new(&mut executor, 20, &tc_account, &mut tc_seqno, false);
+    cancel_burn_with_amount_new(&mut executor, 100, &tc_account, &mut tc_seqno, false);
+}
+
+/// If the DF is upgraded and a DD has a non-zero preburn balance at the time of the
+/// upgrade, make sure that the system doesn't reach a stuck state. But, the DD will need to
+/// upgrade with a nonzero preburn first.
+#[test]
+fn concurrent_preburns_test_upgrade_existing_balance_at_upgrade() {
+    let (mut executor, dr_account, tc_account, dd_account) = start_with_released_df();
+    let mut tc_seqno = 0;
+    let mut dd_seqno = 0;
+    let mut dr_seqno = 1;
+
+    create_preburn_balance(&mut executor, &dd_account, 10, &mut dd_seqno, false);
+
+    // Upgrade the DF
+    upgrade_df(&mut executor, &dr_account, &mut dr_seqno);
+
+    // The new burn/cancel burn scripts will fail since the preburn hasn't been upgraded yet
+    assert!(matches!(
+        cancel_burn_with_amount_new(&mut executor, 100, &tc_account, &mut tc_seqno, true)
+            .unwrap()
+            .status()
+            .status()
+            .unwrap(),
+        KeptVMStatus::MoveAbort(_, 2821)
+    ));
+    assert!(matches!(
+        burn_with_amount_new(&mut executor, 100, &tc_account, &mut tc_seqno, true)
+            .unwrap()
+            .status()
+            .status()
+            .unwrap(),
+        KeptVMStatus::MoveAbort(_, 2821)
+    ));
+
+    // Now upgrade the preburn (recall it already has a nonzero balance).
+    create_preburn_balance(&mut executor, &dd_account, 100, &mut dd_seqno, false);
```

Comment:
@moeziniaf this is the case that the new code is handling - an existing preburn request at the time of the DF upgrade.

Reply:
The code dealing with the df_upgrade, could it be used for testing other upgrade paths? Not sure exactly how all the new test executor stuff works but might we want to document it and put it in a common place for other e2e tests which may need it?

Reply:
Good point! I've updated the PR pulling this functionality out :)